analytic_engine/src/setup.rs

```diff
@@ -158,97 +157,117 @@ pub trait WalsOpener: Send + Sync + Default {
     ) -> Result<OpenedWals>;
 }
 
-/// [RocksEngine] builder.
-#[derive(Default)]
-pub struct RocksDBWalsOpener;
+#[cfg(feature = "wal-rocksdb")]
+pub use rocksdb::*;
+#[cfg(feature = "wal-rocksdb")]
+mod rocksdb {
+    use std::{path::Path, sync::Arc};
+
+    use async_trait::async_trait;
+    use snafu::ResultExt;
+    use table_engine::engine::EngineRuntimes;
+    use wal_rocksdb::manager::Builder as RocksWalBuilder;
+
+    use crate::{
+        setup::{
+            InvalidWalConfig, OpenManifestWal, OpenWal, OpenedWals, Result, WalsOpener,
+            MANIFEST_DIR_NAME, WAL_DIR_NAME,
+        },
+        WalStorageConfig,
+    };
 
-#[async_trait]
-impl WalsOpener for RocksDBWalsOpener {
-    async fn open_wals(
-        &self,
-        config: &WalStorageConfig,
-        engine_runtimes: Arc<EngineRuntimes>,
-    ) -> Result<OpenedWals> {
-        let rocksdb_wal_config = match &config {
-            WalStorageConfig::RocksDB(config) => config.clone(),
-            _ => {
-                return InvalidWalConfig {
-                    msg: format!(
-                        "invalid wal storage config while opening rocksDB wal, config:{config:?}"
-                    ),
+    /// [RocksEngine] builder.
+    #[derive(Default)]
+    pub struct RocksDBWalsOpener;
+
+    #[async_trait]
+    impl WalsOpener for RocksDBWalsOpener {
```

Comment:
This struct should have been moved to `wal-rocksdb` but now it depends on `EngineRuntimes` and `OpenedWals`:

* `OpenedWals`: This can be possibly moved to `wal`;
* `EngineRuntimes`: This comes from table_engine which is tricky to move. But the actually used field is `component/runtime`'s `Runtime` while the certain field is not fixed so we use `engine_runtimes: Arc<EngineRuntimes>` arg here.

Reply:
> This can be possibly moved to wal;

Agree

> But the actually used field is component/runtime's Runtime while the certain field is not fixed so we use engine_runtimes: Arc<EngineRuntimes> arg here.

I think you can only pass Runtime to this function, is there any other issues?

Reply:
> I think you can only pass Runtime to this function, is there any other issues?

You can check the impls:

```rust
// ObkvWalsOpener
let obkv = engine_runtimes
            .write_runtime ...
open_wal_and_manifest_with_table_kv(*obkv_wal_config, engine_runtimes, obkv).await

// MemWalsOpener
open_wal_and_manifest_with_table_kv(
            *obkv_wal_config,
            engine_runtimes,
            self.table_kv.clone(),
        )
        .await

// KafkaWalsOpener
let default_runtime = &engine_runtimes.default_runtime;

// RocksDBWalsOpener
let write_runtime = engine_runtimes.write_runtime.clone();

// ... where open_wal_and_manifest_with_table_kv is
let runtimes = WalRuntimes {
        read_runtime: engine_runtimes.read_runtime.clone(),
        write_runtime: engine_runtimes.write_runtime.clone(),
        default_runtime: engine_runtimes.default_runtime.clone(),
    };
```

So passing a bare `Runtime` may not be suitable for all these variants.

Reply:
Don't expect it get so complicated here...

In theory, all WAL implementation should run where the caller is, but when there are blocking API it's best to spawn on another runtime.

And WAL is a write mainly components, so I guess we only need to pass write runtime to it, other runtime make less sense since WAL are only read on server starts up, no need to pass dedicated read runtime into it.

So maybe we can try combine runtime into one first, if this is possible, then I think this PR could move further.

Reply:
> So maybe we can try combine runtime into one first, if this is possible, then I think this PR could move further

This sounds like the proper way to go. Let me spend some time for investigation.